### PR TITLE
CHINA-117: Add memory profile in log

### PR DESCRIPTION
**Pull request by ianton-ru**
Log memory profile usage
```
2024-03-15 12:32:28,236 MainProcess 1751225 [INFO    ] ch-backup: Command 'backup' completed
2024-03-15 12:32:30,189 MainProcess 1751225 [INFO    ] ch-backup: TraceMalloc top 10 lines
2024-03-15 12:32:30,189 MainProcess 1751225 [INFO    ] ch-backup:   #1: 24997026 B
2024-03-15 12:32:30,190 MainProcess 1751225 [INFO    ] ch-backup:       File "/usr/lib/python3.6/json/decoder.py", line 355
2024-03-15 12:32:30,190 MainProcess 1751225 [INFO    ] ch-backup:         obj, end = self.scan_once(s, idx)
2024-03-15 12:32:30,190 MainProcess 1751225 [INFO    ] ch-backup:   #2: 943184 B
2024-03-15 12:32:30,192 MainProcess 1751225 [INFO    ] ch-backup:       File "/opt/yandex/ch-backup/lib/python3.6/site-packages/ch_backup/util.py", line 86
2024-03-15 12:32:30,192 MainProcess 1751225 [INFO    ] ch-backup:         for file in filter(os.path.isfile, glob.iglob(dir_path + "/**", recursive=True))
2024-03-15 12:32:30,192 MainProcess 1751225 [INFO    ] ch-backup:   #3: 530433 B
2024-03-15 12:32:30,193 MainProcess 1751225 [INFO    ] ch-backup:       File "<frozen importlib._bootstrap_external>", line 487
2024-03-15 12:32:30,193 MainProcess 1751225 [INFO    ] ch-backup:   #4: 406744 B
2024-03-15 12:32:30,194 MainProcess 1751225 [INFO    ] ch-backup:       File "/opt/yandex/ch-backup/lib/python3.6/site-packages/ch_backup/backup/metadata/table_metadata.py", line 72
2024-03-15 12:32:30,194 MainProcess 1751225 [INFO    ] ch-backup:         "disk_name": part.disk_name,
2024-03-15 12:32:30,195 MainProcess 1751225 [INFO    ] ch-backup:   #5: 222741 B
2024-03-15 12:32:30,196 MainProcess 1751225 [INFO    ] ch-backup:       File "/usr/lib/python3.6/multiprocessing/reduction.py", line 51
2024-03-15 12:32:30,196 MainProcess 1751225 [INFO    ] ch-backup:         cls(buf, protocol).dump(obj)
2024-03-15 12:32:30,196 MainProcess 1751225 [INFO    ] ch-backup:   #6: 102004 B
2024-03-15 12:32:30,197 MainProcess 1751225 [INFO    ] ch-backup:       File "/opt/yandex/ch-backup/lib/python3.6/site-packages/botocore/loaders.py", line 276
2024-03-15 12:32:30,197 MainProcess 1751225 [INFO    ] ch-backup:         d for d in os.listdir(possible_path)
2024-03-15 12:32:30,198 MainProcess 1751225 [INFO    ] ch-backup:   #7: 101088 B
2024-03-15 12:32:30,199 MainProcess 1751225 [INFO    ] ch-backup:       File "/opt/yandex/ch-backup/lib/python3.6/site-packages/ch_backup/clickhouse/control.py", line 830
2024-03-15 12:32:30,200 MainProcess 1751225 [INFO    ] ch-backup:         return md5(f.read()).hexdigest()  # nosec
2024-03-15 12:32:30,200 MainProcess 1751225 [INFO    ] ch-backup:   #8: 94502 B
2024-03-15 12:32:30,202 MainProcess 1751225 [INFO    ] ch-backup:       File "/usr/lib/python3.6/json/encoder.py", line 257
2024-03-15 12:32:30,202 MainProcess 1751225 [INFO    ] ch-backup:         return _iterencode(o, 0)
2024-03-15 12:32:30,202 MainProcess 1751225 [INFO    ] ch-backup:   #9: 85128 B
2024-03-15 12:32:30,203 MainProcess 1751225 [INFO    ] ch-backup:       File "/opt/yandex/ch-backup/lib/python3.6/site-packages/botocore/hooks.py", line 485
2024-03-15 12:32:30,203 MainProcess 1751225 [INFO    ] ch-backup:         new_child = {'chunk': part, 'values': None, 'children': {}}
2024-03-15 12:32:30,204 MainProcess 1751225 [INFO    ] ch-backup:   #10: 81280 B
2024-03-15 12:32:30,205 MainProcess 1751225 [INFO    ] ch-backup:       File "/usr/lib/python3.6/glob.py", line 133
2024-03-15 12:32:30,205 MainProcess 1751225 [INFO    ] ch-backup:         names = list(_iterdir(dirname, dironly))
2024-03-15 12:32:30,206 MainProcess 1751225 [INFO    ] ch-backup:     2816 other: 3482185 B
2024-03-15 12:32:30,206 MainProcess 1751225 [INFO    ] ch-backup:     Total allocated size: 31046315 B
```

Turned on by env variable [PYTHONTRACEMALLOC](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONTRACEMALLOC) or by setting `main.tracemalloc` in config file

Final snapshot for main process and periodical (once per 60 seconds) for worker processes.

### CODE REVIEW

File: ch_backup/profile.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        if obj is None:
+            return self.func
+        return partial(self, obj)
+
+    def __call__(self, *args, **kwargs):
+        use_tracemalloc = int(os.environ.get("PYTHONTRACEMALLOC", "0"))
+        if use_tracemalloc > 0:
+            if not tracemalloc.is_tracing():
+                tracemalloc.start()
+        result = self.func(*args, **kwargs)
```
Comment (aalexfvk) [inline ch_backup/profile.py:35]: Let's add `try... finally` to handle possible exceptions from `func`

File: ch_backup/profile.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                logging.info("    {}", line)
+        other = top_stats[self.limit :]
+        if other:
+            size = sum(stat.size for stat in other)
+            logging.info("    {} other: {} B", len(other), size)
+        total = sum(stat.size for stat in top_stats)
+        logging.info("    Total allocated size: {} B", total)
+
+
+def profile(limit, interval=0):
```
Comment (aalexfvk) [inline ch_backup/profile.py:79]: Let's add a default value for `limit` also.

File: ch_backup/profile.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import time
+import tracemalloc
+from functools import partial, wraps
+
+from . import logging
+
+
+class ProfileDecorator:
+    """
+    Class ProfileDecorator memory profiling.
```
Comment (aalexfvk) [inline ch_backup/profile.py:15]: Could you add a short description of `__init__` parameters.

File: ch_backup/profile.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            (
+                tracemalloc.Filter(False, "<frozen importlib._bootstrap>"),
+                tracemalloc.Filter(False, "<unknown>"),
+            )
+        )
+        top_stats = snapshot.statistics("traceback")
+
+        logging.info("TraceMalloc top {} lines", self.limit)
+        for index, stat in enumerate(top_stats[: self.limit], 1):
+            logging.info("  #{}: {} B", index, stat.size)
```
Comment (aalexfvk) [inline ch_backup/profile.py:68]: Let's use `humanfriendly.format_size()` for a size output.

File: ch_backup/profile.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+from . import logging
+
+
+class ProfileDecorator:
+    """
+    Class ProfileDecorator memory profiling.
+    """
+
+    next_snapshot = 0
```
[the remark this answers is not in the document]
Comment (ianton-ru) [reply ch_backup/profile.py:18]: It is exactly for all, i used global variable first, but liner did not allow. Memory snapshot is for all process, not per function, ant this is to write in log not often than once in minute (taking snapshot itself takes near 5-10 seconds).

File: ch_backup/profile.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    def __get__(self, obj, objtype=None):
+        if obj is None:
+            return self.func
+        return partial(self, obj)
+
+    def __call__(self, *args, **kwargs):
+        use_tracemalloc = int(os.environ.get("PYTHONTRACEMALLOC", "0"))
+        if use_tracemalloc > 0:
+            if not tracemalloc.is_tracing():
+                tracemalloc.start()
```
[the remark this answers is not in the document]
Comment (ianton-ru) [reply ch_backup/profile.py:34]: Hm, I thought that it takes nframe value from PYTHONTRACEMALLOC, but not :)

**Review by aalexfvk — APPROVED** [review]

**Pull request merged**